### ISSUE TITLE
Switch.Toggled Triggering & Binding Problems

### ISSUE THREAD

**Issue by antonycartwright** [dotnet/maui#22137]
### Description

Hello all,

I'm unfortunately finding so many simple problems with Maui. I'm not really sure how it's come this far without other people finding these problems. Maybe I'm doing something wrong. Anyway, here's another problem:

A bound switch control triggers the **Toggled** event on load, on manual input and on page closing. This is to be expected as it triggers whenever the toggled value changes, however, there's no way to determine whether the input is user-initiated.

You could have something like this, but it doesn't fully work:

`        Switch sw = (Switch)sender;

        Word w = (Word)((Microsoft.Maui.Controls.Switch)sender).BindingContext;

        if (w != null)
        {
            //The page is not closing, it's either loading or loaded.

            if (!_loaded)
            {
                //Page is loading up, so don't attempt any updates.
            }
            else
            {
                //The page has loaded up, but it may still be populating the switches.
                //For the first x seconds, it's best not to update anything.
                int t = Database.Connection.Update(w);
            }
        }
        else
        {
            //If it's null, then the object has been disposed of (the page is closing), so don't attempt any updates.
        }`

The problem with this is that the listview is **loaded** before all the switches are populated with thei togged state values.

Basically, how do we know when the switch controls have been loaded with the initial values?

1. The switch values are filled with their initial values. Each time the toggled event is triggered.
2. A switch value is modified by a human. Each time the toggled event is triggered.
3. The switch values are emptied on close. Each time the toggled event is triggered.

How are we supposed to deal with this?

### Version with bug

9.0.0-preview.3.10457

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

I was *not* able test on other platforms

**Comment by antonycartwright** [comment]
A working fix/workaround for this is as follows:

Use the Appearing/Disappearing events of the ViewCell to attach/remove the Toggled event from the Switch.

:-)

**Comment by kevinxufei** [comment]
Hi, @antonycartwright .
Do you still have this problem? If the problem still exists, please provide a repo for us to investigate. Thanks!

**Comment by DevelopmentDan** [comment]
This issue is not resolved. Please read the question carefully.

It's easy to reproduce:

Bind the ListView below to an `IEnumerable ` with a ` bool Subscribed` property.
You will see the `SwSubscribed_PropertyChanged` event handler is called for each item that is bound. It is also called if you tap the switch. The problem is, there is no way to tell if it was the tap or the binding that triggered the event. 

```
<ListView x:Name="LvTest"  VerticalOptions="StartAndExpand" HorizontalOptions="FillAndExpand" SelectionMode="None"  >
    <ListView.ItemTemplate>
        <DataTemplate>
            <ViewCell>
                <Grid Padding="5">
                    <Label Text="TEST" VerticalOptions="Center"></Label>
                    <Switch IsToggled="{Binding Subscribed}" PropertyChanged="SwSubscribed_PropertyChanged" ></Switch>
                </Grid>
            </ViewCell>
        </DataTemplate>
    </ListView.ItemTemplate>
</ListView>
```


The triggered event is fired on each Switch regardless of whether the user triggers it, or binding. That wouldn't be an issue, except there is no way to know whether the user triggered the event.

**Comment by antonycartwright** [comment]
> This issue is not resolved. Please read the question carefully.
> 
> It's easy to reproduce:
> 
> Bind the ListView below to an `IEnumerable ` with a ` bool Subscribed` property. You will see the `SwSubscribed_PropertyChanged` event handler is called for each item that is bound. It is also called if you tap the switch. The problem is, there is no way to tell if it was the tap or the binding that triggered the event.
> 
> ```
> <ListView x:Name="LvTest"  VerticalOptions="StartAndExpand" HorizontalOptions="FillAndExpand" SelectionMode="None"  >
>     <ListView.ItemTemplate>
>         <DataTemplate>
>             <ViewCell>
>                 <Grid Padding="5">
>                     <Label Text="TEST" VerticalOptions="Center"></Label>
>                     <Switch IsToggled="{Binding Subscribed}" PropertyChanged="SwSubscribed_PropertyChanged" ></Switch>
>                 </Grid>
>             </ViewCell>
>         </DataTemplate>
>     </ListView.ItemTemplate>
> </ListView>
> ```
> 
> The triggered event is fired on each Switch regardless of whether the user triggers it, or binding. That wouldn't be an issue, except there is no way to know whether the user triggered the event.

Correct. They want me to provide a repo. You don't need a repo to see this. You can literally see it in a few lines! It's a very basic complaint for something that's very primative!

**Comment by DevelopmentDan** [comment]
@antonycartwright I totally agree.

> Use the Appearing/Disappearing events of the ViewCell to attach/remove the Toggled event from the Switch.

Unfortunately, this doesn't work for me - unless I am doing something wrong. Can you give a working example of this work-around?

**Comment by DevelopmentDan** [comment]
@kevinxufei is there any chance of you reopening this issue? The Switch control is pretty useless in a binding scenario unless we are notified whether the user triggered the Toggle event or binding triggered it.

**Comment by DevelopmentDan** [comment]
@antonycartwright there's a possible route to a workaround here:

For each SWITCHED ON switch, during the binding process, this event will be called TWICE. The first time it is called, the `BindingContext ` of the switch will be null. The second time it will contain the object.

When the user clicks the switch, the event is called only ONCE and the `BindingContext` is always present.

I'm trying to work out a way using this difference in behaviour to understand if the Switch is in the process of being bound. Perhaps to stick the switch into an array of switches in a "binding" state (`BindingContext` is null), only to remove them once they are "bound" (`BindingContext` is not null)

```
        private async void SwSubscribed_Toggled(object sender, ToggledEventArgs e)
        {
            Switch sw = (Switch)sender;
            Sub sub = (Sub)sw.BindingContext;
            if (sub != null) await setSubscription(sub.SubId, e.Value);
        }
```

**Comment by DevelopmentDan** [comment]
@antonycartwright I have a simple working work-around for this annoying problem:

In the c# check to see if the value passed in `e.Value` in the Toggled event is different to the value in the `BindingContext`  - if so, we know the user made a change. During binding, `e.Value` will match the value in the `BindingContext` because nothing has changed (and if nothing has changed, then there's nothing to do). But if the user toggles the switch, the value in `e.Value` will no longer match the value in the `BindingContext`, so we know we have something to do.

```
 private async void SwSubscribed_Toggled(object sender, ToggledEventArgs e)
 {
     Switch sw = (Switch)sender;
     Sub sub = (Sub)sw.BindingContext;
     if (sub != null && e.Value != sub.Subscribed) await setSubscription(sub.SubId, e.Value);
 }
```

For this to work, you must set the `Binding` mode to `Mode=OneWay`. This prevents the `Switch` from updating the underlying `object` (accessible in the `BindingContext`) when the switch is toggled. (You can update the `object` yourself if you want to, in the event).

```
<ListView x:Name="LvSub"  VerticalOptions="StartAndExpand" HorizontalOptions="FillAndExpand" SelectionMode="None"  >
    <ListView.ItemTemplate>
        <DataTemplate>
            <ViewCell>
                <Grid Padding="5">
                    <Label Text="{Binding SubName}" VerticalOptions="Center"></Label>
                    <Switch x:Name="SwSubscribed" IsToggled="{Binding Subscribed, Mode=OneWay}" Toggled="SwSubscribed_Toggled" ></Switch>
                </Grid>
            </ViewCell>
        </DataTemplate>
    </ListView.ItemTemplate>
</ListView>
```

**Comment by antonycartwright** [comment]
> @antonycartwright I have a simple working work-around for this annoying problem:
> 
> In the c# check to see if the value passed in `e.Value` in the `Toggled` event is different to the value in the `BindingContext` - if so, we know the user made a change. During binding, `e.Value` will match the value in the `BindingContext` because nothing has changed (and if nothing has changed, then there's nothing to do). But if the user toggles the Switch, the value in `e.Value` will no longer match the value in the `BindingContext`, so we know we have something to do.
> 
> ```
>  private async void SwSubscribed_Toggled(object sender, ToggledEventArgs e)
>  {
>      Switch sw = (Switch)sender;
>      Sub sub = (Sub)sw.BindingContext;
>      if (sub != null && e.Value != sub.Subscribed) await setSubscription(sub.SubId, e.Value);
>  }
> ```
> 
> For this to work, you must set the `Binding` mode to `Mode=OneWay`. This prevents the `Switch` from updating the underlying `object` (accessible in the `BindingContext`) when the Switch is toggled. (You can update the `object` yourself if you want to, in the `Toggled` event).
> 
> ```
> <ListView x:Name="LvSub"  VerticalOptions="StartAndExpand" HorizontalOptions="FillAndExpand" SelectionMode="None"  >
>     <ListView.ItemTemplate>
>         <DataTemplate>
>             <ViewCell>
>                 <Grid Padding="5">
>                     <Label Text="{Binding SubName}" VerticalOptions="Center"></Label>
>                     <Switch x:Name="SwSubscribed" IsToggled="{Binding Subscribed, Mode=OneWay}" Toggled="SwSubscribed_Toggled" ></Switch>
>                 </Grid>
>             </ViewCell>
>         </DataTemplate>
>     </ListView.ItemTemplate>
> </ListView>
> ```

I just add/remove the toggeled event on appear/disappear, but I'll check this out when I get time. Thanks, Antony

**Comment by kevinxufei** [comment]
> @kevinxufei is there any chance of you reopening this issue? The Switch control is pretty useless in a binding scenario unless we are notified whether the user triggered the Toggle event or binding triggered it.

@DevelopmentDan  Sorry, only the author can reopen this issue.

**Comment by kevinxufei** [comment]
> @kevinxufei is there any chance of you reopening this issue? The Switch control is pretty useless in a binding scenario unless we are notified whether the user triggered the Toggle event or binding triggered it.

@DevelopmentDan  Sorry, only the author can reopen this issue.

**Comment by antonycartwright** [comment]
> > @kevinxufei is there any chance of you reopening this issue? The Switch control is pretty useless in a binding scenario unless we are notified whether the user triggered the Toggle event or binding triggered it.
> 
> @DevelopmentDan Sorry, only the author can reopen this issue.

I never closed it - a bot did. The issue is still ongoing as far as I'm aware.
